### PR TITLE
Retrying AlertManager UnaryPath Requests on next replica if one fail

### PR DESCRIPTION
**Pull request by alanprot**
**What this PR does**:

AlertManager unary APIs are failing due a single alert manager fail as distributor picks a random one to forward the request.

This change only pick the replicas alert managers in case of a error.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [X] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -39,6 +39,7 @@
 * [ENHANCEMENT] Memberlist: optimized receive path for processing ring state updates, to help reduce CPU utilization in large clusters. #4345
 * [ENHANCEMENT] Memberlist: expose configuration of memberlist packet compression via `-memberlist.compression=enabled`. #4346
 * [ENHANCEMENT] Updated Prometheus to include changes from prometheus/prometheus#9083. Now whenever `/labels` API calls include matchers, blocks store is queried for `LabelNames` with matchers instead of `Series` calls which was inefficient. #4380
+* [ENHANCEMENT] AlertManager: Retrying AlertManager UnaryPath Requests on next replica if one fail #4422
```
Comment (bboreham) [inline CHANGELOG.md:42]: These descriptions should be understandable by a typical Cortex admin; "UnaryPath" is not in the docs so probably not meaningful.
I'm not familiar with this code but it seems silences, status and receivers are the only unary requests, so could we say that?

File: pkg/alertmanager/distributor.go [hunk 2]
```diff
@@ -254,14 +254,30 @@ func (d *Distributor) doUnary(userID string, w http.ResponseWriter, r *http.Requ
...
+	instances := replicationSet.Instances
+
+	// Randomize the list of instances to not always query the same one.
+	rand.Shuffle(len(instances), func(i, j int) {
+		instances[i], instances[j] = instances[j], instances[i]
+	})
+
+	lastInstance := instances[len(instances)-1]
+
+	for _, instance := range instances {
```
Comment (bboreham) [inline pkg/alertmanager/distributor.go:266]:
```suggestion
	for i, instance := range instances {
```

Then you can write `if err != nil && i == len(instances)-1` and you don't need `lastInstance`.

**Review by bboreham — COMMENTED** [review]
Thanks; looks broadly fine but I made a couple of suggestions.

File: CHANGELOG.md (hunk 1 above)
Comment (alanprot) [reply CHANGELOG.md]: make sense! done!

**Comment by stevesg** [comment]
Thanks for working on this. I can add some context:

The problem with creating silences is that the operation is not idempotent in [upstream] alertmanager. Each API call creates a new silence with a new UUID, which are then replicated to the peers. Adding retries mean you could end up with multiple silences being created from a single API call. e.g. The API call to the first replica is started but doesn't return (e.g. maybe a timeout is hit). That silence is created and the replicated around. If we retry, then the calls to the second/third replica will continue and create duplicates. It's arguable what behavior is better, ideally, we would add an API to upstream alertmanager allowing idempotent silence creation, but this is certainly achievable but non-trivial.

The same problem exists with deletion, it is not idempotent (it returns 4xx if the ID is not found), so you could delete from one replica, try the second, and it has already been deleted due to replication, so you get a 4xx error. This is not such a big problem as for creating

With regards to `/status` and `/receivers`, I don't think we need to retry but can just use a quorum read and pick the first result that comes back (Use the `merger.Noop{}`). In an ideal world the distributor would be transparent, just performing quorum operations with different result merging.

**Comment by alanprot** [comment]
Hi @stevesg 

Yeah.. i was indeed thinking about the Put Silence... I will see if there is something we can do there.

About the delete i think is fine right? If i get a 4xx on the first and retry in the second, it should be fine. 
I can make it not retry on 4xx if needed, but i think should be fine here.

For `/status and /receivers` is still a good thing to do retry it right? Those APIs are returning 5xx on every deployment.

**Comment by stevesg** [comment]
My over-arching point/question was whether we should aim to use quorum operations for everything, so the distributor is logically as transparent as possible (modulo result merging).

> About the delete i think is fine right? If i get a 4xx on the first and retry in the second, it should be fine.
> I can make it not retry on 4xx if needed, but i think should be fine here.

Agreed we can probably make delete work with upstream as-is, preserving the alertmanager API semantic of returning 404 is the annoying bit. Wondering if we should just do quorum write here too? This would make deletes be synchronously replicated. The resulting merging could be along the lines of: if all 404, then 404 else if all 200 or 404 then 200 else error.

> For /status and /receivers is still a good thing to do retry them right? Those APIs are returning 5xx on every deployment.

Absolutely, I was suggesting that you could existing quorum mechanism to save adding complexity.

**Comment by bboreham** [comment]
For adding a silence, could we change the API so that the caller makes up a UUID for the request and passes that in, rather than the alertmanager making one and returning it?

**Comment by alanprot** [comment]
Yeah.. I had a similar idea @bboreham but i was thinking if we could pass an "UUID Generator" on the [Options](https://github.com/prometheus/alertmanager/blob/main/silence/silence.go#L315) and use it [here](https://github.com/prometheus/alertmanager/blob/main/silence/silence.go#L553) - So we could consistently set the same UUID on all the replicas. What do you think?

Besides that, can we change this PR to address the read/delete path and create a new one/issue to address the Put Silence?

**Comment by alanprot** [comment]
Still relevant..

**Comment by alanprot** [comment]
Closing as this is being worked on https://github.com/cortexproject/cortex/pull/4840

**Pull request closed without merging**